File: rbac-manager-for-agents/internal/rbac/whocan_test.go
// Copyright (C) ConfigHub, Inc.
// SPDX-License-Identifier: MIT

package rbac

import (
	"sort"
	"testing"
)

func secretReader() map[string]any {
	return clusterRole("secret-reader", []any{ruleDoc([]string{""}, []string{"secrets"}, []string{"get", "list"})})
}

func TestWhoCan_AcrossClustersWithProvenance(t *testing.T) {
	clusters := build([]FleetResource{
		res("prod-cluster", secretReader()),
		res("prod-cluster", clusterRoleBinding("ops-secrets", "secret-reader", []any{group("ops")})),
		res("dev-cluster", secretReader()),
		res("dev-cluster", clusterRoleBinding("dev-secrets", "secret-reader", []any{group("devs")})),
	})
	grants := WhoCan(clusters, AccessQuery{Verb: "get", Resource: "secrets", APIGroup: ""})
	if len(grants) != 2 {
		t.Fatalf("got %d grants, want 2", len(grants))
	}
	byCluster := map[string]Grant{}
	for _, g := range grants {
		byCluster[g.Cluster] = g
	}
	if byCluster["prod-cluster"].SubjectKey != "Group:ops" {
		t.Errorf("prod subject = %q, want Group:ops", byCluster["prod-cluster"].SubjectKey)
	}
	if byCluster["prod-cluster"].Binding.Name != "ops-secrets" {
		t.Errorf("prod binding = %q, want ops-secrets", byCluster["prod-cluster"].Binding.Name)
	}
	if byCluster["dev-cluster"].SubjectKey != "Group:devs" {
		t.Errorf("dev subject = %q, want Group:devs", byCluster["dev-cluster"].SubjectKey)
	}
}

func TestWhoCan_RoleBindingNamespaceScope(t *testing.T) {
	clusters := build([]FleetResource{
		res("c1", role("ns-reader", "payments", []any{ruleDoc([]string{""}, []string{"secrets"}, []string{"get"})})),
		res("c1", roleBinding("rb", "payments", RoleRef{Kind: "Role", Name: "ns-reader"}, []any{user("alice")})),
	})
	if got := WhoCan(clusters, AccessQuery{Verb: "get", Resource: "secrets", Namespace: "payments"}); len(got) != 1 {
		t.Errorf("payments query = %d grants, want 1", len(got))
	}
	if got := WhoCan(clusters, AccessQuery{Verb: "get", Resource: "secrets", Namespace: "other"}); len(got) != 0 {
		t.Errorf("other query = %d grants, want 0", len(got))
	}
	anywhere := WhoCan(clusters, AccessQuery{Verb: "get", Resource: "secrets"})
	if len(anywhere) != 1 {
		t.Fatalf("anywhere query = %d grants, want 1", len(anywhere))
	}
	if anywhere[0].Scope != "payments" {
		t.Errorf("scope = %q, want payments", anywhere[0].Scope)
	}
}

func TestWhoCan_ClusterAdminBuiltin(t *testing.T) {
	clusters := build([]FleetResource{
		res("c1", clusterRoleBinding("breakglass", "cluster-admin", []any{group("oncall")})),
	})
	grants := WhoCan(clusters, AccessQuery{Verb: "delete", Resource: "secrets", APIGroup: ""})
	if len(grants) != 1 {
		t.Fatalf("got %d grants, want 1", len(grants))
	}
	if !grants[0].ViaBuiltinRole {
		t.Error("expected ViaBuiltinRole=true")
	}
	if grants[0].RoleRefName != "cluster-admin" {
		t.Errorf("roleRefName = %q, want cluster-admin", grants[0].RoleRefName)
	}
}

func TestWhoCan_NoInventedAccessForUnknownBuiltins(t *testing.T) {
	clusters := build([]FleetResource{
		res("c1", clusterRoleBinding("viewers", "view", []any{group("everyone")})),
	})
	if got := WhoCan(clusters, AccessQuery{Verb: "get", Resource: "secrets"}); len(got) != 0 {
		t.Errorf("view binding = %d grants, want 0", len(got))
	}
}

func TestWhoCan_ThroughAggregatedRoles(t *testing.T) {
	clusters := build([]FleetResource{
		res("c1", clusterRole("agg", []any{}, crOpts{
			aggregationRule: map[string]any{"clusterRoleSelectors": []any{
				map[string]any{"matchLabels": map[string]any{"part": "yes"}},
			}},
		})),
		res("c1", clusterRole("part", []any{ruleDoc([]string{""}, []string{"secrets"}, []string{"get"})},
			crOpts{labels: map[string]any{"part": "yes"}})),
		res("c1", clusterRoleBinding("crb", "agg", []any{user("root")})),
	})
	if got := WhoCan(clusters, AccessQuery{Verb: "get", Resource: "secrets"}); len(got) != 1 {
		t.Errorf("aggregated query = %d grants, want 1", len(got))
	}
}

func TestSubjectAccess_AcrossFleet(t *testing.T) {
	clusters := build([]FleetResource{
		res("prod-cluster", secretReader()),
		res("prod-cluster", clusterRoleBinding("b1", "secret-reader", []any{group("ops"), user("alice")})),
		res("dev-cluster", clusterRoleBinding("b2", "cluster-admin", []any{user("alice")})),
	})
	grants := SubjectAccess(clusters, SubjectRef{Kind: "User", Name: "alice"})
	var got []string
	for _, g := range grants {
		got = append(got, g.Cluster+"/"+g.RoleRefName)
	}
	sort.Strings(got)
	want := []string{"dev-cluster/cluster-admin", "prod-cluster/secret-reader"}
	if !equalStrings(got, want) {
		t.Errorf("got %v, want %v", got, want)
	}
}

func TestSubjectAccess_ServiceAccountByNamespace(t *testing.T) {
	clusters := build([]FleetResource{
		res("c1", roleBinding("rb", "apps", RoleRef{Kind: "ClusterRole", Name: "edit"}, []any{sa("ci-deployer", "apps")})),
	})
	if got := SubjectAccess(clusters, SubjectRef{Kind: "ServiceAccount", Name: "ci-deployer", Namespace: "apps"}); len(got) != 1 {
		t.Errorf("apps SA = %d grants, want 1", len(got))
	}
	if got := SubjectAccess(clusters, SubjectRef{Kind: "ServiceAccount", Name: "ci-deployer", Namespace: "other"}); len(got) != 0 {
		t.Errorf("other SA = %d grants, want 0", len(got))
	}
}

func TestAllSubjects_Dedup(t *testing.T) {
	clusters := build([]FleetResource{
		res("c1", clusterRoleBinding("b1", "view", []any{group("devs"), user("alice")})),
		res("c2", clusterRoleBinding("b2", "view", []any{group("devs")})),
	})
	var names []string
	for _, s := range AllSubjects(clusters) {
		names = append(names, s.Name)
	}
	sort.Strings(names)
	if !equalStrings(names, []string{"alice", "devs"}) {
		t.Errorf("got %v, want [alice devs]", names)
	}
}

File: webkit/src/rbac/whocan.test.ts
import { describe, expect, it } from 'vitest';

import {
  build,
  clusterRole,
  clusterRoleBinding,
  group,
  res,
  role,
  roleBinding,
  sa,
  user,
} from './fixtures';
import { allSubjects, subjectAccess, whoCan } from './whocan';

const secretReader = clusterRole('secret-reader', [
  { apiGroups: [''], resources: ['secrets'], verbs: ['get', 'list'] },
]);

describe('whoCan', () => {
  it('finds grants across multiple clusters with provenance', () => {
    const clusters = build([
      res('prod-cluster', secretReader),
      res('prod-cluster', clusterRoleBinding('ops-secrets', 'secret-reader', [group('ops')])),
      res('dev-cluster', secretReader),
      res('dev-cluster', clusterRoleBinding('dev-secrets', 'secret-reader', [group('devs')])),
    ]);
    const grants = whoCan(clusters, { verb: 'get', resource: 'secrets', apiGroup: '' });
    expect(grants).toHaveLength(2);
    const byCluster = Object.fromEntries(grants.map((g) => [g.cluster, g]));
    expect(byCluster['prod-cluster'].subjectKey).toBe('Group:ops');
    expect(byCluster['prod-cluster'].binding.name).toBe('ops-secrets');
    expect(byCluster['dev-cluster'].subjectKey).toBe('Group:devs');
  });

  it('scopes RoleBinding grants to their namespace', () => {
    const clusters = build([
      res('c1', role('ns-reader', 'payments', [
        { apiGroups: [''], resources: ['secrets'], verbs: ['get'] },
      ])),
      res('c1', roleBinding('rb', 'payments', { kind: 'Role', name: 'ns-reader' }, [user('alice')])),
    ]);
    expect(
      whoCan(clusters, { verb: 'get', resource: 'secrets', namespace: 'payments' }),
    ).toHaveLength(1);
    expect(
      whoCan(clusters, { verb: 'get', resource: 'secrets', namespace: 'other' }),
    ).toHaveLength(0);
    // Unscoped query includes namespaced grants, marked with their scope.
    const anywhere = whoCan(clusters, { verb: 'get', resource: 'secrets' });
    expect(anywhere).toHaveLength(1);
    expect(anywhere[0].scope).toBe('payments');
  });

  it('credits cluster-admin builtin bindings with any access', () => {
    const clusters = build([
      res('c1', clusterRoleBinding('breakglass', 'cluster-admin', [group('oncall')])),
    ]);
    const grants = whoCan(clusters, { verb: 'delete', resource: 'secrets', apiGroup: '' });
    expect(grants).toHaveLength(1);
    expect(grants[0].viaBuiltinRole).toBe(true);
    expect(grants[0].roleRefName).toBe('cluster-admin');
  });

  it('does not invent access for unknown builtin roles like view', () => {
    const clusters = build([
      res('c1', clusterRoleBinding('viewers', 'view', [group('everyone')])),
    ]);
    expect(whoCan(clusters, { verb: 'get', resource: 'secrets' })).toHaveLength(0);
  });

  it('answers through aggregated roles', () => {
    const clusters = build([
      res('c1', clusterRole('agg', [], {
        aggregationRule: { clusterRoleSelectors: [{ matchLabels: { part: 'yes' } }] },
      })),
      res('c1', clusterRole('part', [
        { apiGroups: [''], resources: ['secrets'], verbs: ['get'] },
      ], { labels: { part: 'yes' } })),
      res('c1', clusterRoleBinding('crb', 'agg', [user('root')])),
    ]);
    expect(whoCan(clusters, { verb: 'get', resource: 'secrets' })).toHaveLength(1);
  });
});

describe('subjectAccess', () => {
  it('lists every role a subject holds across the fleet', () => {
    const clusters = build([
      res('prod-cluster', secretReader),
      res('prod-cluster', clusterRoleBinding('b1', 'secret-reader', [group('ops'), user('alice')])),
      res('dev-cluster', clusterRoleBinding('b2', 'cluster-admin', [user('alice')])),
    ]);
    const grants = subjectAccess(clusters, { kind: 'User', name: 'alice' });
    expect(grants.map((g) => `${g.cluster}/${g.roleRefName}`).sort()).toEqual([
      'dev-cluster/cluster-admin',
      'prod-cluster/secret-reader',
    ]);
  });

  it('distinguishes ServiceAccounts by namespace', () => {
    const clusters = build([
      res('c1', roleBinding('rb', 'apps', { kind: 'ClusterRole', name: 'edit' }, [
        sa('ci-deployer', 'apps'),
      ])),
    ]);
    expect(
      subjectAccess(clusters, { kind: 'ServiceAccount', name: 'ci-deployer', namespace: 'apps' }),
    ).toHaveLength(1);
    expect(
      subjectAccess(clusters, { kind: 'ServiceAccount', name: 'ci-deployer', namespace: 'other' }),
    ).toHaveLength(0);
  });
});

describe('allSubjects', () => {
  it('dedupes subjects across bindings and clusters', () => {
    const clusters = build([
      res('c1', clusterRoleBinding('b1', 'view', [group('devs'), user('alice')])),
      res('c2', clusterRoleBinding('b2', 'view', [group('devs')])),
    ]);
    expect(allSubjects(clusters).map((s) => s.name).sort()).toEqual(['alice', 'devs']);
  });
});
